compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -827,10 +820,167 @@ impl InitMask {
+    /// The end byte index into `init_mask`.
+    end: Size,
+}
+
+impl<'a> InitChunkIter<'a> {
+    #[inline]
+    fn new(init_mask: &'a InitMask, start: Size, end: Size) -> Self {
+        assert!(start <= end);
+        assert!(end <= init_mask.len);
+
+        let is_init = if start < end { init_mask.get(start) } else { false };
+
+        Self { init_mask, is_init, start, end }
+    }
+}
+
+impl<'a> Iterator for InitChunkIter<'a> {
+    type Item = InitChunk;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.start >= self.end {
+            return None;
+        }
+
+        let end_of_chunk =
+            find_bit(&self.init_mask, self.start, self.end, !self.is_init).unwrap_or(self.end);
+        let range = self.start..end_of_chunk;
+
+        let ret =
+            Some(if self.is_init { InitChunk::Init(range) } else { InitChunk::Uninit(range) });
+
+        self.is_init = !self.is_init;
+        self.start = end_of_chunk;
+
+        ret
+    }
+}
+
+/// Returns the index of the first bit in `start..end` (end-exclusive) that is equal to is_init.
+fn find_bit(init_mask: &InitMask, start: Size, end: Size, is_init: bool) -> Option<Size> {
+    fn find_bit_fast(init_mask: &InitMask, start: Size, end: Size, is_init: bool) -> Option<Size> {
+        fn search_block(
+            bits: Block,
+            block: usize,
+            start_bit: usize,
+            is_init: bool,
+        ) -> Option<Size> {
+            // invert bits so we're always looking for the first set bit
+            let bits = if is_init { bits } else { !bits };
+            // mask off unused start bits
+            let bits = bits & (!0 << start_bit);
+            // find set bit, if any
+            if bits == 0 {
+                None
+            } else {
+                let bit = bits.trailing_zeros();
+                Some(size_from_bit_index(block, bit))
+            }
+        }
+
+        if start >= end {
+            return None;
+        }
+
+        let (start_block, start_bit) = bit_index(start);
+        let (end_block, end_bit) = bit_index(end);
+
+        // handle first block: need to skip `start_bit` bits
+        if let Some(i) =
+            search_block(init_mask.blocks[start_block], start_block, start_bit, is_init)
+        {
+            if i < end {
+                return Some(i);
+            } else {
+                // if the range is less than a block, we may find a matching bit after `end`
+                return None;
+            }
+        }
+
+        let one_block_past_the_end = if end_bit > 0 {
+            // if `end_bit` > 0, then the range overlaps `end_block`
+            end_block + 1
+        } else {
+            end_block
+        };
+
+        // handle remaining blocks
+        if start_block < one_block_past_the_end {
+            for (&bits, block) in init_mask.blocks[start_block + 1..one_block_past_the_end]
+                .iter()
+                .zip(start_block + 1..)
```

Comment:
Is the `zip` here to also get the index?

That might be more clear if you used `enumerate`, though maybe that is less performant?
```rust
init_mask.block.iter()
  .enumerate()
  .skip(start_block + 1)
  .take(...)
```

Reply:
Yeah it's written in a specific way to make the main loop simple, codegen is worse with skip/take (and with various other ways of writing it...I spent far too long trying to make a nice looking and fast implementation). I should've added a comment about this, I will now.